test/files.js

```diff
@@ -0,0 +1,154 @@
+
+let testfile
+let testfileBig
+const testfilePath = path.join(__dirname, './data/testfile.txt')
+
+if (isNode) {
+  const fs = require('fs')
+  testfile = fs.readFileSync(testfilePath)
+  testfileBig = fs.createReadStream(path.join(__dirname, './data/15mb.random'), { bufferSize: 128 })
+} else {
+  testfile = require('raw!./data/testfile.txt')
+}
+
+module.exports = (common) => {
+  describe('.files/add', () => {
+    let ipfs
+
+    before((done) => {
+      common.setup((err, _ipfs) => {
+        expect(err).to.not.exist
+        ipfs = _ipfs
+        done()
+      })
+    })
+
+    after((done) => {
+      common.teardown(done)
+    })
+
+    it('stream', (done) => {
+      const buffered = new Buffer('some data')
+      const rs = new Readable()
+      rs.push(buffered)
+      rs.push(null)
+
+      const arr = []
+      const filePair = {path: 'data.txt', content: rs}
+      arr.push(filePair)
+
+      ipfs.files.add(arr, (err, res) => {
+        expect(err).to.not.exist
+        expect(res).to.be.length(1)
+        expect(res[0].size()).to.equal(17)
+        const mh = 'QmVv4Wz46JaZJeH5PMV4LGbRiiMKEmszPYY3g6fjGnVXBS'
+        expect(bs58.encode(res[0].multihash()).toString()).to.equal(mh)
+        done()
+      })
+    })
+
+    it('buffer as tuple', (done) => {
+      if (!isNode) return done()
+
+      const file = {
+        path: 'testfile.txt',
+        content: new Buffer(testfile)
+      }
+
+      ipfs.files.add([file], (err, res) => {
+        expect(err).to.not.exist
+
+        const added = res[0] != null ? res[0] : res
+        const mh = bs58.encode(added.multihash()).toString()
+        expect(mh).to.equal('Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP')
+        expect(added.links).to.have.length(0)
+        done()
+      })
+    })
+
+    it('buffer', (done) => {
+      let buf = new Buffer(testfile)
+      ipfs.files.add(buf, (err, res) => {
+        expect(err).to.not.exist
+
+        expect(res).to.have.length(1)
+        const mh = bs58.encode(res[0].multihash()).toString()
+        expect(mh).to.equal('Qma4hjFTnCasJ8PVp3mZbZK5g2vGDT4LByLJ7m8ciyRFZP')
+        expect(res[0].links).to.have.length(0)
+        done()
+      })
+    })
+
+    it('BIG buffer', (done) => {
+      if (!isNode) {
+        return done()
+      }
+
+      ipfs.files.add(testfileBig, (err, res) => {
+        expect(err).to.not.exist
+
+        expect(res).to.have.length(1)
+        const mh = bs58.encode(res[0].multihash()).toString()
+        expect(mh).to.equal('Qmcx5werSWQPdrGVap7LARHB4QUSPRPJwxhFuHvdoXqQXT')
+        expect(res[0].links).to.have.length(58)
+        done()
+      })
+    })
+
+    it('add a nested dir as array', (done) => {
+      if (!isNode) return done()
+      const fs = require('fs')
+      const base = path.join(__dirname, 'data/test-folder')
+      const content = (name) => ({
+        path: `test-folder/${name}`,
+        content: fs.readFileSync(path.join(base, name))
+      })
+      const dirs = [
```

Comment:
It would be nice if there was a test for an empty directory by adding one in ```data/test-folder/empty``` so that it can check <a href="https://github.com/ipfs/js-ipfs/pull/256/commits/3b8f4b6448febb0c84aaecb0f111f9001d31b10a">this addition</a> to js-ipfs pr.